File: lua/math/sieve.lua
-- Print the primes below 100, sieved over a table of flags indexed by the number itself.
-- Run: lua sieve.lua

local limit = 100

local is_prime = {}
for n = 2, limit - 1 do
  is_prime[n] = true
end

for n = 2, math.floor(math.sqrt(limit)) do
  if is_prime[n] then
    for multiple = n * n, limit - 1, n do
      is_prime[multiple] = false
    end
  end
end

local primes = {}
for n = 2, limit - 1 do
  if is_prime[n] then
    primes[#primes + 1] = n
  end
end

print(table.concat(primes, " "))

File: kotlin/math/Sieve.kt
// Print the primes below 100, sieved over a BooleanArray whose index is the number itself.
// Run: kotlinc Sieve.kt -include-runtime -d sieve.jar && java -jar sieve.jar

fun main() {
    val limit = 100
    val isPrime = BooleanArray(limit) { it >= 2 }

    var n = 2
    while (n * n < limit) {
        if (isPrime[n]) {
            var multiple = n * n
            while (multiple < limit) {
                isPrime[multiple] = false
                multiple += n
            }
        }
        n++
    }

    println((0 until limit).filter { isPrime[it] }.joinToString(" "))
}
